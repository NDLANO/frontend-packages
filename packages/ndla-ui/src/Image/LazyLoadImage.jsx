/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import PropTypes from 'prop-types';

// Lazyload image with lazysizes (https://github.com/aFarkas/lazysizes)
const LazyLoadImage = ({ alt, src, srcSet, lazyLoadSrc, sizes, ...rest }) => [
  <noscript key="noscript">
    <img alt={alt} src={`${src}`} srcSet={srcSet} sizes={sizes} {...rest} />
  </noscript>,
  <img
    alt={alt}
    key="img"
    style={{ width: '100%' }}
    className="lazyload"
    src={lazyLoadSrc}
    data-src={src}
    data-src-set={srcSet}
    data-sizes={sizes}
    {...rest}
  />,
];

LazyLoadImage.propTypes = {
  alt: PropTypes.string.isRequired,
  src: PropTypes.string.isRequired,
  sizes: PropTypes.string,
  contentType: PropTypes.string,
  srcSet: PropTypes.string,
  lazyLoadSrc: PropTypes.string.isRequired,
};

export default LazyLoadImage;
