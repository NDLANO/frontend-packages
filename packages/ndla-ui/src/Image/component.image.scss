.svg {
  box-shadow: inset 0 0;
}
