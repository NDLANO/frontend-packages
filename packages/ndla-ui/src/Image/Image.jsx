/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import defined from 'defined';
import PropTypes from 'prop-types';
import LazyLoadImage from './LazyLoadImage';

export const makeSrcQueryString = (width, crop, focalPoint) => {
  const cropParams = crop
    ? `&cropStartX=${crop.startX}&cropEndX=${crop.endX}&cropStartY=${crop.startY}&cropEndY=${crop.endY}`
    : '';
  const focalPointParams = focalPoint ? `&focalX=${focalPoint.x}&focalY=${focalPoint.y}` : '';

  return `width=${width}${cropParams}${focalPointParams}`;
};

const getSrcSet = (src, crop, focalPoint) => {
  const widths = [2720, 2080, 1760, 1440, 1120, 1000, 960, 800, 640, 480, 320, 240, 180];
  return widths.map((width) => `${src}?${makeSrcQueryString(width, crop, focalPoint)} ${width}w`).join(', ');
};

const Image = ({ alt, src, lazyLoad, lazyLoadSrc, crop, focalPoint, contentType, sizes, ...rest }) => {
  const srcSet = defined(rest.srcSet, getSrcSet(src, crop, focalPoint));
  const fallbackWidth = defined(rest.fallbackWidth, 1024);
  const queryString = makeSrcQueryString(fallbackWidth, crop, focalPoint);

  if (contentType && contentType === 'image/gif') {
    return <img alt={alt} src={`${src}`} {...rest} />;
  }

  if (lazyLoad) {
    return (
      <LazyLoadImage alt={alt} src={`${src}?${queryString}`} srcSet={srcSet} sizes={sizes} lazyLoadSrc={lazyLoadSrc} />
    );
  }

  return (
    <picture>
      <source type={contentType} srcSet={srcSet} sizes={sizes} />
      <img alt={alt} src={`${src}?${queryString}`} {...rest} />
    </picture>
  );
};

Image.propTypes = {
  alt: PropTypes.string.isRequired,
  src: PropTypes.string.isRequired,
  sizes: PropTypes.string,
  fallbackWidth: PropTypes.number,
  contentType: PropTypes.string,
  srcSet: PropTypes.string,
  lazyLoad: PropTypes.bool,
  lazyLoadSrc: PropTypes.string,
  crop: PropTypes.shape({
    startX: PropTypes.number.isRequired,
    startY: PropTypes.number.isRequired,
    endX: PropTypes.number.isRequired,
    endY: PropTypes.number.isRequired,
  }),
  focalPoint: PropTypes.shape({
    x: PropTypes.number.isRequired,
    y: PropTypes.number.isRequired,
  }),
};

Image.defaultProps = {
  sizes: '(min-width: 1024px) 1024px, 100vw',
};

export default Image;
