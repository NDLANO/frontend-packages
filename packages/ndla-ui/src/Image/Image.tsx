/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React, { ReactNode } from 'react';
import styled from '@emotion/styled';
import LazyLoadImage from './LazyLoadImage';

export interface ImageCrop {
  startX: number;
  startY: number;
  endX: number;
  endY: number;
}

export interface ImageFocalPoint {
  x: number;
  y: number;
}

export const makeSrcQueryString = (width: number | undefined, crop?: ImageCrop, focalPoint?: ImageFocalPoint) => {
  const widthParams = width && `width=${width}`;
  const cropParams =
    crop && `cropStartX=${crop.startX}&cropEndX=${crop.endX}&cropStartY=${crop.startY}&cropEndY=${crop.endY}`;
  const focalPointParams = focalPoint && `focalX=${focalPoint.x}&focalY=${focalPoint.y}`;
  const params = [widthParams, cropParams, focalPointParams].filter((p) => p).join('&');

  return params;
};

const getSrcSet = (src: string, crop: ImageCrop | undefined, focalPoint: ImageFocalPoint | undefined) => {
  const widths = [2720, 2080, 1760, 1440, 1120, 1000, 960, 800, 640, 480, 320, 240, 180];
  return widths.map((width) => `${src}?${makeSrcQueryString(width, crop, focalPoint)} ${width}w`).join(', ');
};

const StyledImageWrapper = styled.div`
  position: relative;
`;
interface Props {
  alt: string;
  src: string;
  sizes?: string;
  fallbackWidth?: number;
  contentType?: string;
  srcSet?: string;
  lazyLoad?: boolean;
  lazyLoadSrc?: string;
  expandButton?: ReactNode;
  crop?: ImageCrop;
  focalPoint?: ImageFocalPoint;
}

const Image = ({
  alt,
  src,
  lazyLoad,
  lazyLoadSrc,
  crop,
  focalPoint,
  contentType,
  sizes = '(min-width: 1024px) 1024px, 100vw',
  expandButton,
  ...rest
}: Props) => {
  const srcSet = rest.srcSet ?? getSrcSet(src, crop, focalPoint);
  const fallbackWidth = rest.fallbackWidth ?? 1024;
  const queryString = makeSrcQueryString(fallbackWidth, crop, focalPoint);

  if (contentType && contentType === 'image/gif') {
    return (
      <StyledImageWrapper>
        <img alt={alt} src={`${src}`} {...rest} />
      </StyledImageWrapper>
    );
  }

  if (lazyLoad) {
    return (
      <StyledImageWrapper>
        <LazyLoadImage
          alt={alt}
          src={`${src}?${queryString}`}
          srcSet={srcSet}
          sizes={sizes}
          lazyLoadSrc={lazyLoadSrc}
        />
      </StyledImageWrapper>
    );
  }

  return (
    <StyledImageWrapper>
      <picture>
        <source type={contentType} srcSet={srcSet} sizes={sizes} />
        <img alt={alt} src={`${src}?${queryString}`} {...rest} />
      </picture>
      {expandButton}
    </StyledImageWrapper>
  );
};

export default Image;
