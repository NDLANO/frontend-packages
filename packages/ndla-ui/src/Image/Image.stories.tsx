/**
 * Copyright (c) 2023-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { Meta, StoryObj } from '@storybook/react';
import Image from './Image';
import { defaultParameters } from '../../../../stories/defaults';

export default {
  title: 'Components/Image',
  tags: ['autodocs'],
  component: Image,
  parameters: {
    inlineStories: true,
    ...defaultParameters,
  },
  args: {
    alt: 'Mann med maske',
    src: 'https://api.test.ndla.no/image-api/raw/id/604',
  },
} as Meta<typeof Image>;

export const Default: StoryObj<typeof Image> = {};

export const Crop: StoryObj<typeof Image> = {
  args: {
    crop: {
      startX: 14.59,
      endX: 79.63,
      startY: 20,
      endY: 100,
    },
  },
};

export const FocalPoint: StoryObj<typeof Image> = {
  args: {
    focalPoint: {
      x: 65.08,
      y: 45.28,
    },
    sizes: '(min-width: 320px) 320px',
  },
};

export const LazyLoad: StoryObj<typeof Image> = {
  args: {
    alt: 'Lysp√¶re',
    src: 'https://api.staging.ndla.no/image-api/raw/Ide.jpg',
    lazyLoad: true,
    lazyLoadSrc: `data:image/svg+xml;utf8,
    <svg xmlns="http://www.w3.org/2000/svg" height="400" width="100%">
      <rect x="0" y="0" width="100%" height="100%" style="fill:#EFF0F2;">
      <animate attributeType="XML" attributeName="fill" values="#EFF0F2;#E8E3E3;#EFF0F2;" dur="3s" repeatCount="indefinite"/>
      </rect>
      </svg>
      `,
  },
};
