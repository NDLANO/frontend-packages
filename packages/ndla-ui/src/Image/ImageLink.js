/**
 * Copyright (c) 2019-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { makeSrcQueryString } from './Image';
import { FocalPointShape, CropShape } from './shapes';

const StyledLink = styled.a`
  box-shadow: inset 0 0;
`;

export function ImageLink({ src, crop, children, ...rest }) {
  return (
    <StyledLink
      target="_blank"
      href={`${src}?${makeSrcQueryString(10720, crop)}`}
      rel="noopener noreferrer"
      {...rest}>
      {children}
    </StyledLink>
  );
}

ImageLink.propTypes = {
  src: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  sizes: PropTypes.string,
  crop: CropShape,
  focalPoint: FocalPointShape,
};
