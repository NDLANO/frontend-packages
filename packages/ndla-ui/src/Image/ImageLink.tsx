/**
 * Copyright (c) 2019-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { ReactNode } from 'react';
import styled from '@emotion/styled';
import { makeSrcQueryString } from './Image';
import { ImageCrop, ImageFocalPoint } from '.';

const StyledLink = styled.a`
  box-shadow: inset 0 0;
`;

interface Props {
  src: string;
  children: ReactNode;
  sizes?: string;
  crop?: ImageCrop;
  focalPoint?: ImageFocalPoint;
}

export function ImageLink({ src, crop, children, ...rest }: Props) {
  return (
    <StyledLink
      target="_blank"
      href={`${src}?${makeSrcQueryString(undefined, crop)}`}
      rel="noopener noreferrer"
      {...rest}
    >
      {children}
    </StyledLink>
  );
}
