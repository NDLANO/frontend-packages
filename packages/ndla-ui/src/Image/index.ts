/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import Image from './Image';
export { makeSrcQueryString } from './Image';
export { ImageLink } from './ImageLink';
export type { ImageCrop, ImageFocalPoint } from './Image';

export default Image;
