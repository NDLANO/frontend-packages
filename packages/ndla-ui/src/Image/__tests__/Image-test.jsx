/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/* eslint-env jest */

import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import Image, { makeSrcQueryString } from '../Image';

test('Image renderers correctly', () => {
  const { container } = render(<Image alt="example" src="https://example.com/image.png" />, { container: null });

  expect(container.firstChild).toMatchSnapshot();
});

test('Lazyloaded image renderers correctly', () => {
  const { container } = render(
    <Image lazyLoad lazyLoadSrc="https://example.com/blurry.png" alt="example" src="https://example.com/image.png" />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('Image with crop and focalpoint props renderers correctly', () => {
  const { container } = render(
    <Image
      crop={{
        startX: 14.59,
        endX: 79.63,
        startY: 20,
        endY: 100,
      }}
      focalPoint={{
        x: 65.08,
        y: 45.28,
      }}
      alt="example"
      src="https://example.com/image.png"
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('makeSrcQueryString renders correctly', () => {
  const crop = {
    startX: 14.59,
    endX: 79.63,
    startY: 20,
    endY: 100,
  };
  const focalPoint = {
    x: 65.08,
    y: 45.28,
  };
  expect(makeSrcQueryString(undefined, undefined, undefined)).toMatch('');
  expect(makeSrcQueryString(1024, undefined, undefined)).toMatch('width=1024');
  expect(makeSrcQueryString(undefined, crop, undefined)).toMatchSnapshot();
  expect(makeSrcQueryString(undefined, undefined, focalPoint)).toMatchSnapshot();
  expect(makeSrcQueryString(1024, crop, focalPoint)).toMatchSnapshot();
});
