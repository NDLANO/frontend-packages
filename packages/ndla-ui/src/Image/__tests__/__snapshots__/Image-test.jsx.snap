// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renderers correctly 1`] = `
.emotion-0 {
  position: relative;
}

.emotion-0 picture {
  width: 100%;
}

.emotion-0[data-svg='true'] {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-0[data-border='true'] {
  border: 1px solid #a5bcd3;
  border-bottom: 0;
  border-radius: 4px;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
}

.emotion-1[data-border='true'] {
  border-radius: 3px;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
}

<div
  class="emotion-0"
  data-svg="false"
>
  <picture>
    <source
      sizes="(min-width: 1024px) 1024px, 100vw"
      srcset="https://example.com/image.png?width=2720 2720w, https://example.com/image.png?width=2080 2080w, https://example.com/image.png?width=1760 1760w, https://example.com/image.png?width=1440 1440w, https://example.com/image.png?width=1120 1120w, https://example.com/image.png?width=1000 1000w, https://example.com/image.png?width=960 960w, https://example.com/image.png?width=800 800w, https://example.com/image.png?width=640 640w, https://example.com/image.png?width=480 480w, https://example.com/image.png?width=320 320w, https://example.com/image.png?width=240 240w, https://example.com/image.png?width=180 180w"
    />
    <img
      alt="example"
      class="emotion-1"
      src="https://example.com/image.png?width=1024"
    />
  </picture>
</div>
`;

exports[`Image with crop and focalpoint props renderers correctly 1`] = `
.emotion-0 {
  position: relative;
}

.emotion-0 picture {
  width: 100%;
}

.emotion-0[data-svg='true'] {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-0[data-border='true'] {
  border: 1px solid #a5bcd3;
  border-bottom: 0;
  border-radius: 4px;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
}

.emotion-1[data-border='true'] {
  border-radius: 3px;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
}

<div
  class="emotion-0"
  data-svg="false"
>
  <picture>
    <source
      sizes="(min-width: 1024px) 1024px, 100vw"
      srcset="https://example.com/image.png?width=2720&cropStartX=14.59&cropEndX=79.63&cropStartY=20&cropEndY=100&focalX=65.08&focalY=45.28 2720w, https://example.com/image.png?width=2080&cropStartX=14.59&cropEndX=79.63&cropStartY=20&cropEndY=100&focalX=65.08&focalY=45.28 2080w, https://example.com/image.png?width=1760&cropStartX=14.59&cropEndX=79.63&cropStartY=20&cropEndY=100&focalX=65.08&focalY=45.28 1760w, https://example.com/image.png?width=1440&cropStartX=14.59&cropEndX=79.63&cropStartY=20&cropEndY=100&focalX=65.08&focalY=45.28 1440w, https://example.com/image.png?width=1120&cropStartX=14.59&cropEndX=79.63&cropStartY=20&cropEndY=100&focalX=65.08&focalY=45.28 1120w, https://example.com/image.png?width=1000&cropStartX=14.59&cropEndX=79.63&cropStartY=20&cropEndY=100&focalX=65.08&focalY=45.28 1000w, https://example.com/image.png?width=960&cropStartX=14.59&cropEndX=79.63&cropStartY=20&cropEndY=100&focalX=65.08&focalY=45.28 960w, https://example.com/image.png?width=800&cropStartX=14.59&cropEndX=79.63&cropStartY=20&cropEndY=100&focalX=65.08&focalY=45.28 800w, https://example.com/image.png?width=640&cropStartX=14.59&cropEndX=79.63&cropStartY=20&cropEndY=100&focalX=65.08&focalY=45.28 640w, https://example.com/image.png?width=480&cropStartX=14.59&cropEndX=79.63&cropStartY=20&cropEndY=100&focalX=65.08&focalY=45.28 480w, https://example.com/image.png?width=320&cropStartX=14.59&cropEndX=79.63&cropStartY=20&cropEndY=100&focalX=65.08&focalY=45.28 320w, https://example.com/image.png?width=240&cropStartX=14.59&cropEndX=79.63&cropStartY=20&cropEndY=100&focalX=65.08&focalY=45.28 240w, https://example.com/image.png?width=180&cropStartX=14.59&cropEndX=79.63&cropStartY=20&cropEndY=100&focalX=65.08&focalY=45.28 180w"
    />
    <img
      alt="example"
      class="emotion-1"
      src="https://example.com/image.png?width=1024&cropStartX=14.59&cropEndX=79.63&cropStartY=20&cropEndY=100&focalX=65.08&focalY=45.28"
    />
  </picture>
</div>
`;

exports[`Lazyloaded image renderers correctly 1`] = `
.emotion-0 {
  position: relative;
}

.emotion-0 picture {
  width: 100%;
}

.emotion-0[data-svg='true'] {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-0[data-border='true'] {
  border: 1px solid #a5bcd3;
  border-bottom: 0;
  border-radius: 4px;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
}

<div
  class="emotion-0"
>
  <noscript />
  <img
    alt="example"
    class="lazyload"
    data-sizes="(min-width: 1024px) 1024px, 100vw"
    data-src="https://example.com/image.png?width=1024"
    data-src-set="https://example.com/image.png?width=2720 2720w, https://example.com/image.png?width=2080 2080w, https://example.com/image.png?width=1760 1760w, https://example.com/image.png?width=1440 1440w, https://example.com/image.png?width=1120 1120w, https://example.com/image.png?width=1000 1000w, https://example.com/image.png?width=960 960w, https://example.com/image.png?width=800 800w, https://example.com/image.png?width=640 640w, https://example.com/image.png?width=480 480w, https://example.com/image.png?width=320 320w, https://example.com/image.png?width=240 240w, https://example.com/image.png?width=180 180w"
    src="https://example.com/blurry.png"
    style="width: 100%;"
  />
</div>
`;

exports[`makeSrcQueryString renders correctly 1`] = `"cropStartX=14.59&cropEndX=79.63&cropStartY=20&cropEndY=100"`;

exports[`makeSrcQueryString renders correctly 2`] = `"focalX=65.08&focalY=45.28"`;

exports[`makeSrcQueryString renders correctly 3`] = `"width=1024&cropStartX=14.59&cropEndX=79.63&cropStartY=20&cropEndY=100&focalX=65.08&focalY=45.28"`;
